New mutator API

timbussmann opened the pull request:
addresses https://github.com/Particular/NServiceBus/issues/4478 by providing a new API based on `EndpointConfiguration` to register mutators instead of having to register them into the container manually.
Under the hood, this still uses the container internally, so this is non-breaking for users currently registering mutators directly via container.

Instead of doing:
```
endpointConfiguration.RegisterComponents(
    registration: components =>
    {
        components.ConfigureComponent<MyMutator>(DependencyLifecycle.InstancePerCall);
    });
```

you can now write:

```
endpointConfiguration.RegisterMessageMutator(new MessageMutator());
```

downside of that API is, that the type param of `RegisterMessageMutator` cannot be limited to the mutator interfaces as we don't have a shared interface between them. So it's currently open to any type and does type checking at invocation time.

Remaining:
* [ ] Update docs to use the new API

andreasohlund commented:
I vote for the more user friendly (IMO) `RegisterMessageMutator` api since you would only look for it when you have built a mutator so I'd don't see users screwing this up

danielmarbach commented:
I'm against type based API because they are bot better than registering on
the container directly. In my opinion it should be either instance based,
func based, list or func<list> Or something similar. The composition root
is available on endpoint level and users can resolve them from the root.

timbussmann commented:
@danielmarbach so you're suggesting something like
```
endpointConfiguration.RegisterMutator(new MyMutator());
// or
endpointConfiguration.RegisterMutator(() => new MyMutator());
```

one thing to consider is, that with the current "API", users can define the container lifetime (e.g. singleton or instance per call). Is that a flexibility we want to keep or are we good with prescribing the lifetime?

timbussmann commented:
btw. with instance based APIs, you can no longer make use of DI by default. Considering the long-term goals, this might be a good reason for instance based APIs?

andreasohlund commented:
👍 for the instance based api

> Is that a flexibility we want to keep or are we good with prescribing the lifetime?

I'm good with us deciding on the lifetime. Same as the behaviors == singleton?

bording commented:
If we go with the instance-based APIs, how do those work with the existing mutator behaviors? We'd need to change the existing behaviors to both build from the container like they currently do, and also add in invoking from the instances/funcs provided by the new API.

> I'm good with us deciding on the lifetime. Same as the behaviors == singleton?

Overall this sounds good, but I do wonder if there are any scenarios for mutators around an assumption of per-call state that would make it hard to migrate to the new registration API.

timbussmann commented:
> If we go with the instance-based APIs, how do those work with the existing mutator behaviors? We'd need to change the existing behaviors to both build from the container like they currently do

as of v6, the existing mutator registrations don't need to be changed. For the next major we can introduce this as a breaking change (that we do no longer support DI and that we limit lifetime). I think it would be easy to bring up a new api allowing for instance per call instantiation if that really is required in case users request that feature?

timbussmann commented:
one thing to consider with the instance based API is, that we allow a single mutator class to imlement multiple mutator interfances, e.g.
```
public class Mutator :
                IMutateIncomingMessages,
                IMutateIncomingTransportMessages,
                IMutateOutgoingMessages,
                IMutateOutgoingTransportMessages
            {...
```

therefore a `RegisterMutator(...)` API with multiple overloads doesn't work nicely. We can either make that API accepting an object or provide different registration methods (e.g. RegisterIncomingMessageMutator, RegisterIncomingTransportMessageMutator, ....). Thoughts?

andreasohlund commented:
Needs a rebase

danielmarbach commented:
Will look into this tomorrow when back

timbussmann commented:
discussed this in the maintainer sync with @dbelcham and @bording and we agreed to move ahead with the instance specific API nevertheless as this seems to be more aligned with future API changes when removing the container. Regarding the issue with the ambiguous parameter, we came up with the following solutions:
* Make the method accept an object and do the type checking at runtime. This isn't really different that what the current proposal does on the type based API, as we also can't limit the type constraints, so no real downside related to my current proposal.
* Introduce a clone of the current mutator interfaces but with a shared marker interface implemented by all specific mutator interfaces. The RegisterMutator API can then be typed to the marker interface. The current interfaces would be obsoleted.

Personally, I'm leaning towards option 1 as I'm not a huge fan of having the same interfaces in different namespaces and the chances of users mixing them up. Thoughts @Particular/nservicebus-maintainers ?

danielmarbach commented:
There is also an option 3:

- Not be too pragmatic about SemVer in this case because we want to give users a good API. So we introduce the marker interface, put it on all the existing mutators.

I think it would be fine in this case. Yes it is a breaking change but since the marker interface doesn't contain any methods a recompile against the version that introduces it would be good enough. From a reflection standpoint is don't see what could break in a real use case. Thoughts?

danielmarbach commented:
Made this WIP again

timbussmann commented:
updated the API and issue description to be instance based. Let me know what you think before I go ahead an raise a doc PR.

danielmarbach commented on src/NServiceBus.Core.Tests/MessageMutators/MutatorRegistrationExtensionsTests.cs:
```diff
@@ -0,0 +1,38 @@
+        }
+
+        [Test]
+        public void Should_not_throw_when_registering_IMutateIncomingMessage()
```
Would we want tests for all the other interfaces as well?

timbussmann commented:
addressed your comments @danielmarbach 

I had to switch the registrations by the extension to address the fact that it would be bound as `object` otherwise as we use `typeof(x)` instead of `x.GetType()` when a specific instance is registered without a specific generic type parameter.

@Particular/container-maintainers do you see an issue with that approach (registering the same instance mutliple times with different interfaces)?

WilliamBZA commented:
@timbussmann we already register multiple interfaces for the same types, the only difference in this case is you are registering a singleton with multiple interfaces. I don't think that will be a problem unless any of the references update the sigleton somewhere. I don't see that happening to the mutator instances - unless they have shared state?

bording commented on src/NServiceBus.Core.Tests/MessageMutators/MutatorRegistrationExtensionsTests.cs:
```diff
@@ -0,0 +1,98 @@
+                exception.Message);
+        }
+
+        [Theory]
```
AFAIK this attribute isn't needed here. I just searched through the code, and I do see several other places where this has been done, but it's not needed. Theories in NUnit are something different than what we're doing. Using `TestCase` by itself should be enough.

timbussmann replied:
yeah I think you're right

bording commented on src/NServiceBus.Core/MessageMutators/MutatorRegistrationExtensions.cs:
```diff
@@ -0,0 +1,58 @@
+
+            if (!registeredMutator)
+            {
+                throw new ArgumentException($"The given instance is no valid message mutator. Implement one of the following mutator interfaces: {typeof(IMutateIncomingMessages).FullName}, {typeof(IMutateIncomingTransportMessages).FullName}, {typeof(IMutateOutgoingMessages).FullName}, {typeof(IMutateOutgoingTransportMessages).FullName}");
```
"The given instance is **not a** valid message mutator."

bording replied:
I'd also suggest an "or" to be added before the final item in the list of interfaces.

bording commented on src/NServiceBus.Core/MessageMutators/Mutators.cs:
```diff
@@ -0,0 +1,19 @@
+
+        protected internal override void Setup(FeatureConfigurationContext context)
+        {
+            context.Pipeline.Register("MutateIncomingTransportMessage", new MutateIncomingTransportMessageBehavior(), "Executes IMutateIncomingTransportMessages");
```
By moving the incoming mutator registrations out of `ReceiveFeature`, I was wondering how that would impact SendOnly endpoints, but then I realized that `ReceiveFeature` runs even when and endpoint is SendOnly. I suppose that means there's no problem here, but now I wonder why `ReceiveFeature` isn't set up with a prerequisite around SendOnly.

timbussmann replied:
![](https://imgflip.com/s/meme/Ancient-Aliens.jpg)

timbussmann replied:
(In general it would definitely make sense to move the mutators into their own feature, but I think that's out of scope here?)

bording commented on src/NServiceBus.Core/MessageMutators/MutatorRegistrationExtensions.cs:
```diff
@@ -0,0 +1,58 @@
+            var registeredMutator = false;
+            foreach (var mutatorInterface in GetImplementedMutatorInterfaces(messageMutator))
+            {
+                // register the mutator with it's specific mutator interface as it will be registered as object instead
```
"its" not "it's"

timbussmann replied:
constantly messing those up, thanks.

bording commented on src/NServiceBus.Core/MessageMutators/MutatorRegistrationExtensions.cs:
```diff
@@ -0,0 +1,58 @@
+            foreach (var mutatorInterface in GetImplementedMutatorInterfaces(messageMutator))
+            {
+                // register the mutator with it's specific mutator interface as it will be registered as object instead
+                endpointConfiguration.RegisterComponents(c => c.RegisterSingleton(mutatorInterface, messageMutator));
```
I have to say that the name and xml comments for `RegisterSingleton` are rather confusing. At first I was thinking this would mean we'd only be able to register a single mutator for each interface type! Had to run a quick test to verify that this wasn't the case.

timbussmann replied:
I tried to clarify this a bit more, although I think this would be an edge case rather than the default approach. Have a second look.

timbussmann commented:
@bording addressed your comments. Please have another look
